### PR TITLE
updates serialized graph lookup for newly added languages #10079

### PR DESCRIPTION
**Pull request by chrabyrd**
### Types of changes

-   [x] Bugfix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Issues Solved

#10079

### CODE REVIEW

File: arches/app/models/models.py [hunk 1]
```diff
@@ -520,16 +520,20 @@ def is_editable(self):
...
         try:
             graph = PublishedGraph.objects.get(publication=self.publication, language=language)
         except PublishedGraph.DoesNotExist:
-            if raise_if_missing:
-                raise
-            graph = None
+            try:
+                from arches.app.utils.i18n import LanguageSynchronizer  # prevents circular import
+
+                LanguageSynchronizer.synchronize_settings_with_db()
```
Comment (jacobtylerwalls) [inline arches/app/models/models.py:533]: I retested #10079, and when using the language switcher, the page just hangs with a spinner (no response is received from Django). This was with dockerized AfS.

I added a print statement before the call to `synchronize...` , including the graph name, and got:

```py
2024-03-05 14:18:22 syncing... 1 Instrument
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:22 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Arches System Settings
2024-03-05 14:18:23 syncing... 1 Attribute Assignment 
2024-03-05 14:18:23 syncing... 1 Name
2024-03-05 14:18:23 syncing... 1 Dimension
2024-03-05 14:18:23 syncing... 1 Monetary Amount
2024-03-05 14:18:23 syncing... 1 TimeSpan
2024-03-05 14:18:24 syncing... 1 __Activity_Template
2024-03-05 14:18:24 syncing... 1 Statement
2024-03-05 14:18:24 syncing... 1 Identifier
2024-03-05 14:18:24 syncing... 1 Instrument
2024-03-05 14:18:24 syncing... 1 Collection or Set
2024-03-05 14:18:24 syncing... 1 Project
2024-03-05 14:18:24 syncing... 1 Observation
2024-03-05 14:18:25 syncing... 1 Physical Thing
2024-03-05 14:18:25 syncing... 1 Person
2024-03-05 14:18:26 syncing... 1 Place
2024-03-05 14:18:26 syncing... 1 Textual Work
2024-03-05 14:18:26 syncing... 1 Sampling Activity
2024-03-05 14:18:26 syncing... 1 Modification
2024-03-05 14:18:27 syncing... 1 Digital Resources
```

Comment (jacobtylerwalls) [reply arches/app/models/models.py:533]: Looks like pg was choking:
```
2024-03-05 14:25:41 2024-03-05 19:25:41.146 PST [1706] LOG:  unexpected EOF on client connection with an open transaction
```

**Comment by chrabyrd** [comment]
@jacobtylerwalls Yeah it look like the change I made was a bit presumptuous, as calling `LanguageSynchronizer` updated _all_ graphs with the new language, which could cause unexpected effects in the frontend. 

Instead of further mucking around in i18n logic by either updating the `synchronize_settings_with_database` to take an optional graphid arg or creating a method that largely mimics the functionality but only for a single graph ( and therefore _some_ graphs could have _some_ translations -- bleh ) I opted to just have the logic fallback to not using a serialized graph if it can't find one. Tested on AfS (non-dockerized) with Russian and everything looks to be working.

This also shouldn't regress #10189 any longer, as that looked to be solved by having the datatype fetch logic in `Tile.save` match the logic match the pattern in `Tile.delete`

**Comment by jacobtylerwalls** [comment]
> This also shouldn't regress https://github.com/archesproject/arches/issues/10189 any longer, as that looked to be solved by having the datatype fetch logic in Tile.save match the logic match the pattern in Tile.delete

With the latest I'm still getting the error described in #10189 with the testing instructions [here](https://github.com/archesproject/arches/pull/10203#issue-1965957506):

```sh
python3 manage.py packages -o import_business_data -s /usr/local/afs-test-pkg/pkg/business_data/Group.csv -ow True
```
```py
2024-03-18 13:30:02,705 arches.app.utils.data_management.resources.formats.csvfile WARNING  2024-03-18 13:30:02.671172: WARNING: failed to save document in resource: 90e76649-d9ed-493f-88c3-92e4e5d2b77a on line 5. Exception detail:
'NoneType' object is not subscriptable
```

**Comment by jacobtylerwalls** [comment]
Also, just I retested #10079 and got this 500 error leaving the UI with a hang:

```sh
2024-03-18 09:40:43 Internal Server Error: /api/bulk_disambiguated_resource_instance
2024-03-18 09:40:43 Traceback (most recent call last):
2024-03-18 09:40:43   File "/usr/local/lib/python3.10/dist-packages/django/core/handlers/exception.py", line 55, in inner
2024-03-18 09:40:43     response = get_response(request)
2024-03-18 09:40:43   File "/usr/local/lib/python3.10/dist-packages/django/core/handlers/base.py", line 197, in _get_response
2024-03-18 09:40:43     response = wrapped_callback(request, *callback_args, **callback_kwargs)
2024-03-18 09:40:43   File "/usr/local/lib/python3.10/dist-packages/django/views/generic/base.py", line 104, in view
2024-03-18 09:40:43     return self.dispatch(request, *args, **kwargs)
2024-03-18 09:40:43   File "/web_root/arches/arches/app/views/api.py", line 107, in dispatch
2024-03-18 09:40:43     return super(APIBase, self).dispatch(request, *args, **kwargs)
2024-03-18 09:40:43   File "/usr/local/lib/python3.10/dist-packages/django/views/generic/base.py", line 143, in dispatch
2024-03-18 09:40:43     return handler(request, *args, **kwargs)
2024-03-18 09:40:43   File "/web_root/arches/arches/app/views/api.py", line 1319, in get
2024-03-18 09:40:43     disambiguated_resource_instances[str(resource.pk)] = resource.to_json(
2024-03-18 09:40:43   File "/web_root/arches/arches/app/models/resource.py", line 783, in to_json
2024-03-18 09:40:43     return LabelBasedGraphV2.from_resource(
2024-03-18 09:40:43   File "/web_root/arches/arches/app/utils/label_based_graph_v2.py", line 217, in from_resource
2024-03-18 09:40:43     published_graph = models.PublishedGraph.objects.get(publication=resource.graph.publication, language=user_language)
2024-03-18 09:40:43   File "/usr/local/lib/python3.10/dist-packages/django/db/models/manager.py", line 87, in manager_method
2024-03-18 09:40:43     return getattr(self.get_queryset(), name)(*args, **kwargs)
2024-03-18 09:40:43   File "/usr/local/lib/python3.10/dist-packages/django/db/models/query.py", line 637, in get
2024-03-18 09:40:43     raise self.model.DoesNotExist(
2024-03-18 09:40:43 arches.app.models.models.PublishedGraph.DoesNotExist: PublishedGraph matching query does not exist.
2024-03-18 09:40:43 [18/Mar/2024 13:40:43] "GET /api/bulk_disambiguated_resource_instance?v=beta&resource_ids=5498a6e4-e895-46a8-a3a6-a584ea0162df&uncompacted=true HTTP/1.1" 500 139323
```

**Comment by jacobtylerwalls** [comment]
Looks like it's the same error as #9852

File: arches/app/utils/index_database.py [hunk 2]
```diff
@@ -41,7 +41,10 @@ def get_serialized_graph(graph):
 
     if graph.graphid not in serialized_graphs:
         published_graph = graph.get_published_graph()
-        serialized_graphs[graph.graphid] = published_graph.serialized_graph
+
+        if published_graph:
+            serialized_graphs[graph.graphid] = published_graph.serialized_graph
+
```
Comment (chiatt) [inline arches/app/utils/index_database.py:47]: If published_graph is false it looks like you're going to get a key error when the function tries to return `serialized_graphs[graph.graphid]`

**Comment by jacobtylerwalls** [comment]
As this fixes a bug that raises KeyError, I think we should target 7.5.

**Comment by chrabyrd** [comment]
This has been proven to be quite complex, and is being closed in favor of #11342

**Pull request closed without merging**